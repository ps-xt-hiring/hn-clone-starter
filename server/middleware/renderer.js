import React from 'react';
import ReactDOMServer from 'react-dom/server';
import Loadable from 'react-loadable';
import { Provider as ReduxProvider } from 'react-redux';
import { StaticRouter } from 'react-router';
import { Helmet } from 'react-helmet';
import { ServerStyleSheets } from '@material-ui/styles';
// import our main App component
import App from '../../src/App';

const path = require('path');
const fs = require('fs');

// import the manifest generated with the create-react-app build
const manifestFile = path.resolve(
  __dirname,
  '..',
  '..',
  'build',
  'asset-manifest.json',
);
const manifest = JSON.parse(fs.readFileSync(manifestFile, 'utf8'));

// function to extract js assets from the manifest
const extractAssets = (assets, chunks) => Object.keys(assets)
  .filter(asset => chunks.indexOf(asset.replace('.js', '')) > -1)
  .map(k => assets[k]);

const extractAssetsCss = (assets, chunks) => Object.keys(assets)
  .filter(asset => chunks.indexOf(asset.replace('.css', '')) > -1)
  .map(k => assets[k]);

export default store => (req, res) => {
  // get the html file created with the create-react-app build
  const filePath = path.resolve(__dirname, '..', '..', 'build', 'index.html');
  const sheets = new ServerStyleSheets();
  fs.readFile(filePath, 'utf8', (err, htmlData) => {
    if (err) {
      console.error('err', err);
      return res.status(404).end();
    }

    const modules = [];
    const routerContext = {};

    // render the app as a string
    const html = ReactDOMServer.renderToString(
      sheets.collect(
        <Loadable.Capture report={m => modules.push(m)}>
          <ReduxProvider store={store}>
            <StaticRouter location={req.baseUrl} context={routerContext}>
              <App />
            </StaticRouter>
          </ReduxProvider>
        </Loadable.Capture>,
      ),
    );

    // get the stringified state
    const reduxState = JSON.stringify(store.getState());

    // map required assets to script tags
    const extraChunks = extractAssets(manifest, modules).map(
      c => `<script type="text/javascript" src="${c}"></script>`,
    );

    const extraChunksCss = extractAssetsCss(manifest, modules).map(
      c => `<link rel="stylesheet" href="${c}" />`,
    );

    const css = sheets.toString();

    // get HTML headers
    const helmet = Helmet.renderStatic();

    // now inject the rendered app into our html and send it to the client
    return res.send(
      htmlData
        // write the React app
        .replace('<div id="root"></div>', `<div id="root">${html}</div>`)
        .replace(
          '<style id="css-server-side"></style>',
          `<style id="css-server-side">${css}</style>`,
        )
        // write the string version of our state
        .replace('__REDUX_STATE__={}', `__REDUX_STATE__=${reduxState}`)
        // append the extra css assets
        .replace('</head>', `${extraChunksCss.join('')}</head>`)
        // append the extra js assets
        .replace('</body>', `${extraChunks.join('')}</body>`)
        // write the HTML header tags
        .replace(
          '<title></title>',
          helmet.title.toString() + helmet.meta.toString(),
        ),
    );
  });
};
