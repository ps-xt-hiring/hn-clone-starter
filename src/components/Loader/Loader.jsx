import React from 'react';
import { LOADING } from '../../utils/constants';

export default function Loader({ loading }) {
  return (
    <>
      {
        loading
        && <p className="loading">{ LOADING }</p>
      }
    </>
  );
}
