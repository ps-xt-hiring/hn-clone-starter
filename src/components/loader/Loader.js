import React from 'react';
import NProgress from 'nprogress';
import Router from 'next/router';
import { PropTypes } from 'prop-types';
import loaderCss from './Loader.css';

class Loader extends React.Component {
  constructor(props) {
    super(props);
    this.timer = null;
  }

  componentDidMount() {
    const { options } = this.props;

    if (options) {
      NProgress.configure(options);
    }

    Router.events.on('routeChangeStart', this.routeChangeStart);
    Router.events.on('routeChangeComplete', this.routeChangeEnd);
    Router.events.on('routeChangeError', this.routeChangeEnd);
  }

  routeChangeStart = () => {
    const { startPosition } = this.props;
    NProgress.set(startPosition);
    NProgress.start();
  };

  routeChangeEnd = () => {
    clearTimeout(this.timer);
    const { stopDelayMs } = this.props;
    this.timer = setTimeout(() => {
      NProgress.done(true);
    }, stopDelayMs);
  };

  render() {
    return (
      <style jsx global>
        {`${loaderCss}`}
      </style>
    );
  }
}

Loader.defaultProps = {
  startPosition: 0.3,
  stopDelayMs: 200,
  options: {},
};

Loader.propTypes = {
  startPosition: PropTypes.number,
  stopDelayMs: PropTypes.number,
  options: PropTypes.shape({}),
};

export default Loader;
