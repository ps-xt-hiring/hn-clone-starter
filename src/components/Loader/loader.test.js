import React from 'react';
import renderer from 'react-test-renderer';
import Loader from './Loader';

describe('Loader Component', () => {
  it('Should have loading parameter', () => {
    const component = renderer.create(<Loader loading={null} />);
    const componentInstance = component.root;
    expect(Object.prototype.hasOwnProperty.call(componentInstance.props, 'loading')).toBeTruthy();
  });

  it('Should not return any ouput on false', () => {
    const component = renderer.create(<Loader loading={false} />);
    const componentInstance = component.root;
    expect(componentInstance.props.loading).toBe(false);
    expect(component).toMatchSnapshot();
  });

  it('Should return on true', () => {
    const component = renderer.create(<Loader loading />);
    const componentInstance = component.root;
    expect(componentInstance.props.loading).toBe(true);
    expect(component).toMatchSnapshot();
  });
});
