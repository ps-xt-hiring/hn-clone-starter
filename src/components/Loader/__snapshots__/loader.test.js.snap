// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loader Component Should not return any ouput on false 1`] = `null`;

exports[`Loader Component Should return on true 1`] = `
<p
  className="loading"
>
  Loading...
</p>
`;
